src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -647,18 +673,51 @@ object LazyList extends LazyListFactory[LazyList] {
       this
     }
 
-    protected def tailDefined: Boolean = tlEvaluated
-    protected def headDefined: Boolean = hdEvaluated
+    protected[immutable] def tailDefined: Boolean = tlEvaluated
+    protected[immutable] def headDefined: Boolean = hdEvaluated
   }
 
+  @SerialVersionUID(3L)
+  final class Suspended[A](var eval: () => LazyList[A]) extends LazyList[A] {
+    private[this] var evaluated: Boolean = false
+    lazy val next: LazyList[A] = {
+      evaluated = true
+      var res = eval()
+      eval = null
+      while (res.isInstanceOf[Suspended[_]]) {
+        // skip through multiple suspensions in a loop rather than using the stack
+        res = res.asInstanceOf[Suspended[A]].next
```

Comment:
Is there a test to make sure this is stack-safe?

Reply:
There will be.

- [ ] test for stack safety

Reply:
I'm actually fairly certain this isn't stack-safe, because each call to `next` checks `isInstanceOf[Suspended[_]]`, and then calls `next` on the result. Thus, you get a call stack (of sorts) like the one below

```
    res.res.res.res.next
 at res.res.res.next
 at res.res.next
 at res.next
 at this.next
```